### ISSUE TITLE
Linux XFCE submenu empty

### ISSUE THREAD

**Issue by mrpalide** [fyne-io/systray#12]
Hey there!
We have an app that use systray, and now I want to submit this report as issue.
On Gnome works like a charm. But XFCE is unstable in submenu!, and KDE Plasma is really strange and useless on submenus.
You can see bellow screencasts:

#### Ubuntu
OS: Ubuntu 22.04 (Minimal Installation)
Desktop: Gnome 42.0

https://user-images.githubusercontent.com/79150699/167568719-9211a4c6-570d-43c5-8d81-73499bd43209.mp4

#### Nitrux (Debianbased)
OS: Nitrux 2.1.1
Desktop: KDE Plasma 5.24.4

https://user-images.githubusercontent.com/79150699/167568652-ef396dcb-4a39-42ca-b193-424b9e290699.mov

#### Manjaro (Archbased)
OS: Manjaro 21.2.6
Desktop: XFCE

https://user-images.githubusercontent.com/79150699/167568666-b2287651-70f4-4d41-9368-17e8e373f59a.mov

**Comment by andydotxyz** [comment]
@mrpalide can you please try the fixes on https://github.com/fyne-io/systray/pull/11

**Comment by mrpalide** [comment]
> @mrpalide can you please try the fixes on #11

I test on #11 actually.

**Comment by mrpalide** [comment]
I found a huge problem actually on example app!
About 100 data race actually appear on that simple example!
First built it with `-race` flag as `go build -race example/main.go`, then run it `./main`.

Use `Quit` button on systray app to see this on console:
```
Quit2 now...
Exit at 2022-05-11 13:15:54.215513558 +0430 +0430 m=+4.429888999
Found 100 data race(s)
```

Also this is a shot of submenu problem of example app on XFCE desktop:
![image](https://user-images.githubusercontent.com/79150699/167809320-893a465a-c7f6-4c66-86f1-f9b1cd846938.png)

**Comment by andydotxyz** [comment]
> Also this is a shot of submenu problem of example app on XFCE desktop:

Can you please confirm which version of XFCE? I cannot replicate the problem with 4.16.
Be sure to also test the latest `master` of the systray repo :)

**Comment by mrpalide** [comment]
Test, and same. See this video:
https://user-images.githubusercontent.com/79150699/168971085-2740fcaf-dc39-439a-a648-13142f1204ab.mov

**Comment by andydotxyz** [comment]
Thank you for this.
Can you please confirm if you have any plugins on or configuration regarding the panel?
I will try to build a minimal arch to see if I can replicate in the same OS, but so far I can't make it not work in this way.

**Comment by mrpalide** [comment]
Hey andy, sorry for late response.
Yes, I can confirm that no plugin or configuration on panel added.

**Comment by slytomcat** [comment]
`Also this is a shot of submenu problem of example app on XFCE desktop:`
I found such problem not only in Status Notification Plugin (that serves D-Bus org.kde.StatusNotifierWatcher in XFCE) but also in Indicator plugin (can be used via GTK+ AppIndicator API). 
So, it seems (for me at least) that this problem is somewheree deeper in GTK, but not in the panel plugins.

**Comment by slytomcat** [comment]
About data races - I fixed a lot of them in [my fork](https://github.com/slytomcat/systray).
It was a big problem for my project yd-go where menu is used very actively (it's updated rather often).

**Comment by slytomcat** [comment]
I also can confirm that KDE notification tray plugin always show submenu on wrong side (partly outside the screen) during the first opening of submenu. I see it in my tests in Kubuntu 22.04 (fresh install without and with latest updates).

**Comment by andydotxyz** [comment]
> About data races - I fixed a lot of them in [my fork](https://github.com/slytomcat/systray).

Great news, if you could open some pull requests I am sure they would be gladly received :).

**Comment by andydotxyz** [comment]
> And from my point of view it is also not a systray issue but it's an issue of KDE notification tray plugin or problem is somewhere deeper.

Agreed, we don't control how the menu will be rendered so can't fix that particular issue I think

**Comment by slytomcat** [comment]
Ok. I'll do the PR with my fixes.

**Comment by andydotxyz** [comment]
Thank you so much @slytomcat - let's see what they say.
Seems to be some activity so thanks for kicking this off https://gitlab.xfce.org/xfce/xfce4-panel/-/issues/582

**Comment by slytomcat** [comment]
That's why I love the Open Source World: you 'almost always' :) can find people who can help you.

**Comment by slytomcat** [comment]
https://gitlab.xfce.org/xfce/xfce4-panel/-/issues/582 is closed as wrong working lib is not maintained any more :(

There is my simply workaround (deployed into my yd-go project):

```bash
PATH_TO_PATCH="$(go env GOMODCACHE)/$(cat go.mod | grep 'github.com/godbus/dbus/v5' | sed 's/\s*\(\S\+\) \(.*\)$/\1@\2/')"
chmod a+w $PATH_TO_PATCH
patch <conn.patch -f $PATH_TO_PATCH/conn.go
```
while conn.patch file contains:
```
@@ -432,7 +432,7 @@
 		case TypeSignal:
 			conn.handleSignal(sequence, msg)
 		case TypeMethodCall:
-			go conn.handleCall(msg)
+			conn.handleCall(msg)
 		}
 
 	}
```
Patching code can be added to build script like I done it [here](https://github.com/slytomcat/yd-go/blob/master/build.sh)

Unfortunately I see no better solution at the moment. 

BTW: there is some plans to move XFCE panel back-end to Ayatana indicator project. But Ayatana has another big issue: [It dosn't  support IconPixmap property of org.freedesktop.StatusNotifierItem interface](https://github.com/AyatanaIndicators/libayatana-appindicator/issues/47).

**Comment by slytomcat** [comment]
https://github.com/godbus/dbus/issues/327 - there is already one opened and closed later about this.

The patch is just a simplest workaround but not the fix of real issue (in [libdbusmenu](https://github.com/AyatanaIndicators/libdbusmenu))

**Comment by andydotxyz** [comment]
Ah yes I see, thanks for all of this.
Sadly I'm not sure that we can do anything more in this ticket then, perhaps should just close it - the docs you provide and workaround leave a nice trail for others to find

**Comment by mrpalide** [comment]
So we should hide sub-menus on XFCE desktops, right? :)

**Comment by slytomcat** [comment]
@mrpalide no, you can apply the patch to the godbus for your build and sub-menus on XFCE will work fine (because your app will handle the requests from libdbusmenu in the sequense as they comes). It's not perfect solution, but it works.

**Comment by slytomcat** [comment]
C is not my favorite language but I'll try to find the problem in that lib. 

Probably there is someone more experienced in C who can find and fix this issue in libdbusmenu.
